src/lib/rng/rng.cpp

```diff
@@ -9,35 +9,44 @@
 #include <botan/internal/loadstor.h>
 #include <botan/internal/os_utils.h>
 
+#if defined(BOTAN_HAS_SYSTEM_RNG)
+  #include <botan/system_rng.h>
+#endif
+
 namespace Botan {
 
-void RandomNumberGenerator::randomize_with_ts_input(uint8_t output[], size_t output_len)
+void RandomNumberGenerator::randomize_with_ts_input(std::span<uint8_t> output)
    {
    if(this->accepts_input())
       {
-      /*
-      Form additional input which is provided to the PRNG implementation
-      to paramaterize the KDF output.
-      */
-      uint8_t additional_input[16] = { 0 };
-      store_le(OS::get_system_timestamp_ns(), additional_input);
-      store_le(OS::get_high_resolution_clock(), additional_input + 8);
+      constexpr auto s_hd_clk = sizeof(decltype(OS::get_high_resolution_clock()));
+      constexpr auto s_sys_ts = sizeof(decltype(OS::get_system_timestamp_ns()));
+      constexpr auto s_pid    = sizeof(decltype(OS::get_process_id()));
+
+      std::array<uint8_t, s_hd_clk + s_sys_ts + s_pid> additional_input = {0};
+      auto s_additional_input = std::span(additional_input.begin(), additional_input.end());
+
+      store_le(OS::get_high_resolution_clock(), s_additional_input.data());
+      s_additional_input = s_additional_input.subspan(s_hd_clk);
+
+#if defined(BOTAN_HAS_SYSTEM_RNG)
+      System_RNG system_rng;
+      system_rng.randomize(s_additional_input);
+#else
+      store_le(OS::get_system_timestamp_ns(), s_additional_input.data());
+      s_additional_input = s_additional_input.subspan(s_sys_ts);
 
-      this->randomize_with_input(output, output_len, additional_input, sizeof(additional_input));
+      store_le(OS::get_process_id(), s_additional_input.data());
+#endif
+
+      this->fill_bytes_with_input(output, additional_input);
       }
    else
       {
-      this->randomize(output, output_len);
+      this->fill_bytes_with_input(output, {});
       }
    }
```

Comment:
/cc @randombit

The base implementation of `RandomNumberGenerator::randomize_with_ts_input()` now adopted the behavior of `Stateful_RNG` (as discussed [here](https://github.com/randombit/botan/pull/3195#issuecomment-1449602320) and [here](https://github.com/randombit/botan/pull/3195#issuecomment-1477692624)). Additionally, this method is not virtual any longer: I.e. all RNGs will inherit exactly this implementation.

There's one caveat: Naturally, also the `System_RNG` is a subclass of `RandomNumberGenerator`. In other words: if the platform's RNG reports that it can accept input, it ends up seeding itself. Technically, that shouldn't be problematic though its worth noting.

There shouldn't be a chance of this becoming an endless loop either. `system_rng.randomize()` will resolve to `System_RNG::fill_bytes_with_input()` (without actual input). By design, this is the most low-level method that will not call back into higher-level members (like `randomize_with_ts_input()` to actually create an endless recursion).